src/main/java/io/github/thebusybiscuit/slimefun4/utils/InfiniteBlockGenerator.java

```diff
@@ -0,0 +1,177 @@
+     */
+    @Override
+    public boolean test(@Nonnull Block b) {
+        Validate.notNull(b, "Block cannot be null!");
+
+        /*
+         * This will eliminate non-matching base materials If we
+         * are on a version without Basalt, it will be null here and not match.
+         */
+        if (b.getType() == getGeneratedMaterial()) {
+            switch (this) {
+            case COBBLESTONE_GENERATOR:
+                return hasSurroundingMaterials(b, Material.WATER, Material.LAVA);
+            case STONE_GENERATOR:
+                if (b.getRelative(BlockFace.UP).getType() == Material.LAVA) {
+                    return hasSurroundingMaterials(b, Material.WATER);
+                } else {
+                    return false;
+                }
+            case BASALT_GENERATOR:
+                if (b.getRelative(BlockFace.DOWN).getType() == Material.SOUL_SOIL) {
+                    return hasSurroundingMaterials(b, Material.LAVA, Material.BLUE_ICE);
+                } else {
+                    return false;
+                }
+            default:
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
+    @ParametersAreNonnullByDefault
+    private boolean hasSurroundingMaterials(Block b, Material... materials) {
+        Validate.notNull(b, "The Block cannot be null!");
+        Validate.notEmpty(materials, "Materials need to have a size of at least one!");
+
+        boolean[] matches = new boolean[materials.length];
+        int count = 0;
+
+        for (BlockFace face : sameLevelFaces) {
+            Block neighbour = b.getRelative(face);
+            Material neighbourType = neighbour.getType();
+
+            for (int i = 0; i < materials.length; i++) {
+                if (neighbourType == materials[i] && !matches[i]) {
+                    matches[i] = true;
+                    count++;
+                    break;
+                }
+            }
+
+            if (count == materials.length) {
+                return true;
+            }
+        }
+
+        return count == materials.length;
+    }
+
+    /**
+     * This method calls a {@link BlockFormEvent} for this {@link InfiniteBlockGenerator}.
+     * There are a few plugins who catch these events to inject custom {@link Material Materials} into
+     * Cobblestone Generators, so we wanna give them the oppurtunity to catch this as well.
+     * 
+     * @param from
+     *            The {@link Block} where our liquid is coming from.
+     * @param to
+     *            The {@link Block} our liquid has flown to / solidified at.
+     * 
+     * @return Our called {@link BlockFormEvent}
+     */
+    @Nonnull
+    public BlockFormEvent callEvent(@Nonnull Block block) {
+        BlockState state = PaperLib.getBlockState(block, false).getState();
+        BlockFormEvent event = new BlockFormEvent(block, state);
+        Bukkit.getPluginManager().callEvent(event);
+        return event;
+    }
```

Comment:
Public method so Validate#notNull here? Or am I still getting this wrong xD

Reply:
```suggestion
    public BlockFormEvent callEvent(@Nonnull Block block) {
        Validate.notNull(block, "The Block cannot be null!");
        BlockState state = PaperLib.getBlockState(block, false).getState();
        BlockFormEvent event = new BlockFormEvent(block, state);
        Bukkit.getPluginManager().callEvent(event);
        return event;
    }
```